File: src/Comm/TCP_IP/seer_api/test/test_transport.py
"""seer_api.transport 회귀 시험.

핵심 고정 대상:
  1. 공식 SDK `packMsg` 와 **바이트 동일** — 우리가 만든 프레임이 로봇이 받는 프레임과 같은가.
  2. 부분 수신을 끝까지 모으는가(공식 데모의 recv(1024) 절단 결함을 재도입하지 않는가).
  3. 응답 편호·seq 를 실제로 대조하는가(이관 전 두 구현은 둘 다 대조하지 않았다).
  4. 요청 간 최소 간격을 지키는가.

시험은 실기 없이 돈다 — 소켓을 가짜로 갈아끼운다.
"""
import importlib.util
import json
import os
import struct
import sys

import pytest

sys.path.insert(0, os.path.join(os.path.dirname(__file__), ".."))

from seer_api import ports, transport  # noqa: E402
from seer_api.transport import (  # noqa: E402
    SeerProtocolError,
    SeerTransport,
    pack,
    unpack_head,
)

# ---- 공식 SDK 원본 로드 (원문 대조용) ----
_SDK_PATH = os.path.join(
    os.path.dirname(__file__),
    "../../../../../References/Seer-Driver/github_sdk/Robokit_TCP_API_py/netprotocol/rbkNetProtoEnums.py",
)


def _load_official():
    path = os.path.abspath(_SDK_PATH)
    if not os.path.exists(path):
        pytest.skip(f"공식 SDK 원본 없음: {path}")
    spec = importlib.util.spec_from_file_location("rbkNetProtoEnums_ref", path)
    mod = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(mod)
    return mod


class FakeSocket:
    """프로그램된 바이트를 조금씩 흘려주는 소켓 대역.

    recv 를 1~3B 씩 쪼개 돌려주어 부분 수신 경로를 반드시 지나게 한다.
    """

    def __init__(self, response=b"", chunk=3):
        self.response = response
        self.chunk = chunk
        self.sent = b""
        self.closed = False
        self.connected_to = None
        self.timeout = None
        self._pos = 0

    def settimeout(self, t):
        self.timeout = t

    def connect(self, addr):
        self.connected_to = addr

    def sendall(self, data):
        self.sent += data

    def recv(self, n):
        take = min(n, self.chunk, len(self.response) - self._pos)
        if take <= 0:
            return b""
        out = self.response[self._pos:self._pos + take]
        self._pos += take
        return out

    def close(self):
        self.closed = True


def make_response(seq, api_type, payload: dict, sync=transport.SYNC):
    body = json.dumps(payload).encode("ascii")
    head = struct.pack(transport.HEAD_FMT, sync, transport.VERSION, seq,
                       len(body), api_type, transport.RSV)
    return head + body


def bind(monkeypatch, sock):
    monkeypatch.setattr(transport.socket, "socket", lambda *a, **k: sock)


# ---------- 1. 공식 SDK 바이트 동일성 ----------

def test_pack_matches_official_sdk_with_body():
    """본문 있는 요청이 공식 packMsg 와 바이트 동일."""
    official = _load_official()
    msg = {"x": 10.0, "y": 3.0, "angle": 0}
    assert pack(7, 2002, msg) == official.packMsg(7, 2002, msg)


def test_pack_matches_official_sdk_empty_body():
    """빈 요청은 body 없이 length=0 — 공식 원본의 `if(msg != {})` 분기와 같아야 한다."""
    official = _load_official()
    for api in (1000, 1004, 1009, 1050):
        assert pack(1, api, {}) == official.packMsg(1, api, {})
        assert pack(1, api, None) == official.packMsg(1, api, {})


def test_head_format_string_matches_official():
    official = _load_official()
    assert transport.HEAD_FMT == official.PACK_HEAD_FMT_STR
    assert transport.RSV == official.PACK_RSV_DATA
    assert transport.HEAD_LEN == 16


def test_pack_header_fields_are_literal():
    """헤더 각 바이트가 사양대로인지 직접 확인(포맷 문자열에 의존하지 않는 독립 검증)."""
    frame = pack(0x1234, 1009, None)
    assert len(frame) == 16
    assert frame[0] == 0x5A
    assert frame[1] == 0x01
    assert frame[2:4] == b"\x12\x34"  # seq, big-endian
    assert frame[4:8] == b"\x00\x00\x00\x00"  # length 0
    assert frame[8:10] == struct.pack(">H", 1009)
    assert frame[10:16] == b"\x00" * 6


# ---------- 2. 수신 ----------

def test_recv_exact_reassembles_chunked_response(monkeypatch):
    payload = {"lasers": [{"device_info": {"device_name": "front"}}], "ret_code": 0}
    sock = FakeSocket(make_response(1, 11009, payload), chunk=1)  # 1바이트씩
    bind(monkeypatch, sock)
    tr = SeerTransport("1.2.3.4", ports.API_PORT_STATE, min_interval=0)
    assert tr.request(1009) == payload


def test_large_response_is_not_truncated(monkeypatch):
    """공식 데모의 recv(1024) 절단 결함이 재도입되지 않았는지."""
    payload = {"beams": list(range(2000)), "ret_code": 0}
    sock = FakeSocket(make_response(1, 11009, payload), chunk=997)
    bind(monkeypatch, sock)
    tr = SeerTransport("1.2.3.4", ports.API_PORT_STATE, min_interval=0)
    got = tr.request(1009)
    assert len(got["beams"]) == 2000


def test_connection_closed_midway_raises(monkeypatch):
    full = make_response(1, 11009, {"ret_code": 0})
    sock = FakeSocket(full[:10])  # 헤더도 못 채우고 끊김
    bind(monkeypatch, sock)
    tr = SeerTransport("1.2.3.4", ports.API_PORT_STATE, min_interval=0)
    with pytest.raises(ConnectionError):
        tr.request(1009)
    assert sock.closed, "끊긴 소켓을 남기면 다음 요청이 재연결하지 못한다"


# ---------- 3. 대조 ----------

def test_unexpected_response_type_raises(monkeypatch):
    sock = FakeSocket(make_response(1, 60000, {}))  # 엉뚱한 포트로 보내면 60000
    bind(monkeypatch, sock)
    tr = SeerTransport("1.2.3.4", ports.API_PORT_STATE, min_interval=0)
    with pytest.raises(SeerProtocolError, match="60000"):
        tr.request(1009)


def test_seq_mismatch_raises(monkeypatch):
    sock = FakeSocket(make_response(999, 11009, {"ret_code": 0}))  # seq 반향 안 함
    bind(monkeypatch, sock)
    tr = SeerTransport("1.2.3.4", ports.API_PORT_STATE, min_interval=0)
    with pytest.raises(SeerProtocolError, match="seq"):
        tr.request(1009)


def test_bad_sync_byte_raises(monkeypatch):
    sock = FakeSocket(make_response(1, 11009, {}, sync=0x00))
    bind(monkeypatch, sock)
    tr = SeerTransport("1.2.3.4", ports.API_PORT_STATE, min_interval=0)
    with pytest.raises(SeerProtocolError, match="sync"):
        tr.request(1009)


def test_connection_limit_rejection_is_its_own_error(monkeypatch):
    """한도 거부는 "편호 불일치"가 아니라 전용 예외로 나와야 한다.

    **[실측 2026-08-07]** 19204 에 9번째 연결 시 로봇이 보낸 실제 프레임을 그대로 재현한다 —
    편호가 요청+10000 이 아니라 **포트 번호(19204)** 이고 `ret_code` 가 61001 이다.
    이 서명을 못 알아보면 `응답 편호 19204 (기대 11004)` 라는 오해를 부르는 진단이 나온다.
    """
    body = (b'{"create_on":"2026-08-07T18:36:40.679+0900",'
            b'"err_msg":"reach the maximum of status api connection limitation",'
            b'"ip":"192.168.44.2","port":52248,"ret_code":61001}')
    head = struct.pack(transport.HEAD_FMT, transport.SYNC, transport.VERSION, 1,
                       len(body), 19204, transport.RSV)
    sock = FakeSocket(head + body)
    bind(monkeypatch, sock)
    tr = SeerTransport("1.2.3.4", ports.API_PORT_STATE, min_interval=0)
    with pytest.raises(transport.SeerConnectionLimitError, match="61001"):
        tr.request(1004)


def test_connection_limit_error_is_a_protocol_error():
    """기존 `except SeerProtocolError` 호출자를 깨지 않는다."""
    assert issubclass(transport.SeerConnectionLimitError, SeerProtocolError)


def test_other_type_mismatch_still_generic(monkeypatch):
    """포트 번호가 아닌 편호 불일치는 종전대로 일반 예외 — 한도 예외로 뭉뚱그리지 않는다."""
    sock = FakeSocket(make_response(1, 60000, {"ret_code": 61001}))
    bind(monkeypatch, sock)
    tr = SeerTransport("1.2.3.4", ports.API_PORT_STATE, min_interval=0)
    with pytest.raises(SeerProtocolError) as exc:
        tr.request(1004)
    assert not isinstance(exc.value, transport.SeerConnectionLimitError)


def test_unpack_head_rejects_short_buffer():
    with pytest.raises(SeerProtocolError, match="헤더 길이"):
        unpack_head(b"\x5a\x01\x00")


def test_seq_increments_between_requests(monkeypatch):
    """seq 고정(이관 전 두 구현의 결함)이 아니라 순환하는지."""
    seqs = []

    class SeqSocket(FakeSocket):
        def sendall(self, data):
            seqs.append(struct.unpack(">H", data[2:4])[0])
            self._pos = 0
            self.response = make_response(seqs[-1], 11004, {"ret_code": 0})

    sock = SeqSocket(b"")
    bind(monkeypatch, sock)
    tr = SeerTransport("1.2.3.4", ports.API_PORT_STATE, min_interval=0)
    for _ in range(3):
        tr.request(1004)
    assert seqs == [1, 2, 3]


def test_expect_type_override(monkeypatch):
    """4011 처럼 응답 편호를 명시해야 하는 경우."""
    sock = FakeSocket(make_response(1, 14011, {"ret_code": 0}))
    bind(monkeypatch, sock)
    tr = SeerTransport("1.2.3.4", ports.API_PORT_STATE, min_interval=0)
    assert tr.request(4011, expect_type=14011) == {"ret_code": 0}


# ---------- 4. 스로틀 ----------

def test_min_interval_is_enforced(monkeypatch):
    now = [0.0]
    slept = []

    class SeqSocket(FakeSocket):
        def sendall(self, data):
            self._pos = 0
            self.response = make_response(struct.unpack(">H", data[2:4])[0], 11004, {})

    sock = SeqSocket(b"")
    bind(monkeypatch, sock)
    tr = SeerTransport(
        "1.2.3.4", ports.API_PORT_STATE, min_interval=0.1,
        clock=lambda: now[0], sleep=lambda s: (slept.append(s), now.__setitem__(0, now[0] + s)),
    )
    tr.request(1004)          # 첫 요청 — 대기 없음
    assert slept == []
    tr.request(1004)          # 즉시 재요청 — 0.1s 대기해야 한다
    assert slept and abs(slept[0] - 0.1) < 1e-9
    now[0] += 5.0
    tr.request(1004)          # 충분히 지난 뒤 — 추가 대기 없음
    assert len(slept) == 1


def test_zero_min_interval_never_sleeps(monkeypatch):
    slept = []

    class SeqSocket(FakeSocket):
        def sendall(self, data):
            self._pos = 0
            self.response = make_response(struct.unpack(">H", data[2:4])[0], 11004, {})

    sock = SeqSocket(b"")
    bind(monkeypatch, sock)
    tr = SeerTransport("1.2.3.4", ports.API_PORT_STATE, min_interval=0,
                       sleep=lambda s: slept.append(s))
    tr.request(1004)
    tr.request(1004)
    assert slept == []


# ---------- 5. 연결 수명 ----------

def test_context_manager_connects_and_closes(monkeypatch):
    sock = FakeSocket(b"")
    bind(monkeypatch, sock)
    with SeerTransport("9.9.9.9", 19204) as tr:
        assert tr.is_connected
        assert sock.connected_to == ("9.9.9.9", 19204)
    assert sock.closed and not tr.is_connected


def test_connect_is_idempotent(monkeypatch):
    made = []

    def factory(*a, **k):
        s = FakeSocket(b"")
        made.append(s)
        return s

    monkeypatch.setattr(transport.socket, "socket", factory)
    tr = SeerTransport("1.2.3.4", 19204)
    tr.connect()
    tr.connect()
    assert len(made) == 1

File: src/Comm/seer_tcp_ip/test/test_transport.cpp
// 전송 계층 회귀 시험.
//
// 고정 대상:
//   1. 공식 SDK packMsg 와 **바이트 동일** — 우리가 만든 프레임이 로봇이 받는 프레임과 같은가.
//   2. 부분 수신을 끝까지 모으는가(공식 데모의 recv(1024) 절단 결함을 재도입하지 않는가).
//   3. 응답 편호·seq 를 실제로 대조하는가.
//   4. 요청 간 최소 간격을 지키는가.
//   5. 한도 거부(61001)를 전용 예외로 구분하는가.
#include <cstring>
#include <string>
#include <vector>

#include "fake_stream.hpp"
#include "harness.hpp"
#include "seer_tcp_ip/transport.hpp"

using namespace seer_tcp_ip;

namespace
{

std::string hex(const std::vector<char> &v, std::size_t n)
{
    static const char *d = "0123456789ABCDEF";
    std::string s;
    for (std::size_t i = 0; i < n && i < v.size(); ++i)
    {
        const unsigned char c = static_cast<unsigned char>(v[i]);
        s += d[c >> 4];
        s += d[c & 0xF];
    }
    return s;
}

struct FakeClock
{
    std::int64_t now = 0;
    std::vector<std::int64_t> slept;
    ClockMs clock() { return [this] { return now; }; }
    SleepMs sleep()
    {
        return [this](std::int64_t ms) {
            slept.push_back(ms);
            now += ms;
        };
    }
};

}  // namespace

int main()
{
    // ---------- 1. 프레임 생성 ----------
    {
        CASE("무파라미터 프레임은 16B, 헤더 각 바이트가 사양대로");
        const std::vector<char> f = pack(0x1234, 1009);
        CHECK_EQ(f.size(), std::size_t(16));
        //  5A   01  |  1234  |  00000000  |  03F1  |  000000000000
        //  sync ver    seq        len=0        1009     예약 6B
        CHECK_EQ(hex(f, 16), std::string("5A0112340000000003F1000000000000"));
    }
    {
        CASE("본문이 있으면 길이 필드가 본문 바이트 수와 같다");
        const Json msg{{"x", 10.0}};
        const std::vector<char> f = pack(7, 2002, msg);
        const std::string body = msg.dump();
        CHECK_EQ(f.size(), 16 + body.size());
        const Head h = unpackHead(f.data(), 16);
        CHECK_EQ(h.jsonLen, static_cast<std::uint32_t>(body.size()));
        CHECK_EQ(h.apiType, std::uint16_t(2002));
        CHECK_EQ(h.seq, std::uint16_t(7));
    }
    {
        CASE("빈 객체는 본문을 붙이지 않는다 — 공식 packMsg 의 if(msg != {}) 분기");
        for (std::uint16_t api : {std::uint16_t(1000), std::uint16_t(1004), std::uint16_t(1050)})
        {
            CHECK_EQ(pack(1, api, Json::object()).size(), std::size_t(16));
            CHECK_EQ(pack(1, api).size(), std::size_t(16));
        }
    }
    {
        CASE("예약 6바이트가 0 으로 채워진다");
        const std::vector<char> f = pack(1, 1000);
        for (int i = 10; i < 16; ++i)
        {
            CHECK_EQ(f[i], char(0));
        }
    }

    // ---------- 2. 헤더 해석 ----------
    {
        CASE("짧은 버퍼 거부");
        const char buf[3] = {0x5A, 0x01, 0x00};
        CHECK_THROWS_MSG(unpackHead(buf, 3), ProtocolError, "헤더 길이");
    }
    {
        CASE("sync 불일치 거부");
        std::string r = fake::makeResponse(1, 11009, "{}", 0x00);
        CHECK_THROWS_MSG(unpackHead(r.data(), 16), ProtocolError, "sync");
    }

    // ---------- 3. 수신 ----------
    {
        CASE("1바이트씩 와도 끝까지 모은다");
        fake::Stream s;
        s.chunk = 1;
        s.reset(fake::makeResponse(1, 11009, R"({"ret_code":0,"lasers":[1,2,3]})"));
        Transport tr("1.2.3.4", ports::kState, 1.0, 0, fake::factoryFor(&s));
        const Json r = tr.request(1009);
        CHECK_EQ(r["lasers"].size(), std::size_t(3));
    }
    {
        CASE("큰 응답이 잘리지 않는다 — recv(1024) 절단 결함 재도입 금지");
        std::string big = "{\"ret_code\":0,\"beams\":[";
        for (int i = 0; i < 2000; ++i)
        {
            big += (i ? "," : "") + std::to_string(i);
        }
        big += "]}";
        fake::Stream s;
        s.chunk = 997;
        s.reset(fake::makeResponse(1, 11009, big));
        Transport tr("1.2.3.4", ports::kState, 1.0, 0, fake::factoryFor(&s));
        CHECK_EQ(tr.request(1009)["beams"].size(), std::size_t(2000));
    }
    {
        CASE("수신 중 끊기면 예외 + 소켓 정리");
        fake::Stream s;
        const std::string full = fake::makeResponse(1, 11009, "{}");
        s.reset(full.substr(0, 10));  // 헤더도 못 채우고 끊김
        Transport tr("1.2.3.4", ports::kState, 1.0, 0, fake::factoryFor(&s));
        CHECK_THROWS_MSG(tr.request(1009), TransportError, "수신 중 연결 종료");
        CHECK(!tr.isConnected());
    }

    // ---------- 4. 대조 ----------
    {
        CASE("엉뚱한 응답 편호 거부");
        fake::Stream s;
        s.reset(fake::makeResponse(1, 60000, "{}"));
        Transport tr("1.2.3.4", ports::kState, 1.0, 0, fake::factoryFor(&s));
        CHECK_THROWS_MSG(tr.request(1009), ProtocolError, "60000");
    }
    {
        CASE("seq 반향 불일치 거부");
        fake::Stream s;
        s.reset(fake::makeResponse(999, 11009, "{}"));
        Transport tr("1.2.3.4", ports::kState, 1.0, 0, fake::factoryFor(&s));
        CHECK_THROWS_MSG(tr.request(1009), ProtocolError, "seq");
    }
    {
        CASE("seq 는 고정이 아니라 순환한다");
        fake::Stream s;
        std::vector<std::uint16_t> seqs;
        s.onWrite = [&seqs](fake::Stream &st, const std::vector<char> &f) {
            const Head h = unpackHead(f.data(), 16);
            seqs.push_back(h.seq);
            st.reset(fake::makeResponse(h.seq, 11004, "{}"));
        };
        Transport tr("1.2.3.4", ports::kState, 1.0, 0, fake::factoryFor(&s));
        for (int i = 0; i < 3; ++i)
        {
            tr.request(1004);
        }
        const std::vector<std::uint16_t> want = {1, 2, 3};
        CHECK_EQ(seqs, want);
    }
    {
        CASE("expectType 지정(4011 → 14011)");
        fake::Stream s;
        s.reset(fake::makeResponse(1, 14011, "{}"));
        Transport tr("1.2.3.4", ports::kState, 1.0, 0, fake::factoryFor(&s));
        CHECK_NOTHROW(tr.request(4011, Json::object(), 14011));
    }

    // ---------- 5. 한도 거부 ----------
    {
        CASE("한도 거부는 전용 예외 — 편호가 포트 번호로 온다");
        fake::Stream s;
        s.reset(fake::makeResponse(
            1, ports::kState,
            R"({"ret_code":61001,"err_msg":"reach the maximum of status api connection limitation"})"));
        Transport tr("1.2.3.4", ports::kState, 1.0, 0, fake::factoryFor(&s));
        CHECK_THROWS_MSG(tr.request(1004), ConnectionLimitError, "61001");
    }
    {
        CASE("한도 예외는 ProtocolError 의 하위 — 기존 호출자를 안 깬다");
        fake::Stream s;
        s.reset(fake::makeResponse(1, ports::kState, R"({"ret_code":61001,"err_msg":"x"})"));
        Transport tr("1.2.3.4", ports::kState, 1.0, 0, fake::factoryFor(&s));
        CHECK_THROWS_MSG(tr.request(1004), ProtocolError, "61001");
    }
    {
        CASE("포트 번호가 아닌 편호 불일치는 일반 예외로 남는다");
        fake::Stream s;
        s.reset(fake::makeResponse(1, 60000, R"({"ret_code":61001})"));
        Transport tr("1.2.3.4", ports::kState, 1.0, 0, fake::factoryFor(&s));
        bool wasLimit = false;
        try
        {
            tr.request(1004);
        }
        catch (const ConnectionLimitError &)
        {
            wasLimit = true;
        }
        catch (const ProtocolError &)
        {
        }
        CHECK(!wasLimit);
    }

    // ---------- 6. 스로틀 ----------
    {
        CASE("요청 간 최소 간격을 지킨다");
        fake::Stream s;
        s.onWrite = [](fake::Stream &st, const std::vector<char> &f) {
            const Head h = unpackHead(f.data(), 16);
            st.reset(fake::makeResponse(h.seq, 11004, "{}"));
        };
        FakeClock fc;
        Transport tr("1.2.3.4", ports::kState, 1.0, 100, fake::factoryFor(&s), fc.clock(),
                     fc.sleep());
        tr.request(1004);
        CHECK(fc.slept.empty());     // 첫 요청은 대기 없음
        tr.request(1004);            // 즉시 재요청 → 100ms 대기
        CHECK_EQ(fc.slept.size(), std::size_t(1));
        CHECK_EQ(fc.slept[0], std::int64_t(100));
        fc.now += 5000;
        tr.request(1004);            // 충분히 지난 뒤 → 추가 대기 없음
        CHECK_EQ(fc.slept.size(), std::size_t(1));
    }
    {
        CASE("간격 0 이면 대기하지 않는다");
        fake::Stream s;
        s.onWrite = [](fake::Stream &st, const std::vector<char> &f) {
            const Head h = unpackHead(f.data(), 16);
            st.reset(fake::makeResponse(h.seq, 11004, "{}"));
        };
        FakeClock fc;
        Transport tr("1.2.3.4", ports::kState, 1.0, 0, fake::factoryFor(&s), fc.clock(), fc.sleep());
        tr.request(1004);
        tr.request(1004);
        CHECK(fc.slept.empty());
    }

    // ---------- 7. 연결 수명 ----------
    {
        CASE("연결 실패는 죽은 스트림을 남기지 않는다 — 다음 요청이 새로 연결한다");
        // 실기 실패경로에서 드러난 결함: 실패한 connect 뒤 stream_ 이 남아 다음 요청이
        // 죽은 서술자에 send 를 걸고 `Bad file descriptor` 로 한 번 더 헛돌았다.
        struct Flaky : IByteStream
        {
            int *attempts;
            bool *failNext;
            std::string body;
            std::size_t pos = 0;
            bool open = false;
            void connect(const std::string &, std::uint16_t, double) override
            {
                ++(*attempts);
                if (*failNext)
                {
                    throw TransportError("connect 실패(시험)");
                }
                open = true;
            }
            std::size_t read(char *b, std::size_t n) override
            {
                const std::size_t take = std::min(n, body.size() - pos);
                if (take == 0) { return 0; }
                std::memcpy(b, body.data() + pos, take);
                pos += take;
                return take;
            }
            void writeAll(const char *, std::size_t) override
            {
                if (!open) { throw TransportError("Bad file descriptor"); }
            }
            void close() override { open = false; }
            bool isOpen() const override { return open; }
        };
        int attempts = 0;
        bool failNext = true;
        auto factory = [&attempts, &failNext]() -> std::unique_ptr<IByteStream> {
            auto f = std::unique_ptr<Flaky>(new Flaky());
            f->attempts = &attempts;
            f->failNext = &failNext;
            f->body = fake::makeResponse(1, 11004, "{}");
            return f;
        };
        Transport tr("1.2.3.4", ports::kState, 1.0, 0, factory);
        CHECK_THROWS_MSG(tr.request(1004), TransportError, "connect 실패");
        CHECK(!tr.isConnected());          // 죽은 스트림이 남으면 안 된다
        failNext = false;
        CHECK_NOTHROW(tr.request(1004));   // 다음 요청은 새로 연결해 성공한다
        CHECK_EQ(attempts, 2);
    }
    {
        CASE("connect 는 멱등, close 후 재연결");
        fake::Stream s;
        s.reset(fake::makeResponse(1, 11004, "{}"));
        Transport tr("9.9.9.9", ports::kState, 1.0, 0, fake::factoryFor(&s));
        tr.connect();
        tr.connect();
        CHECK(tr.isConnected());
        CHECK_EQ(s.connectedIp, std::string("9.9.9.9"));
        CHECK_EQ(s.connectedPort, ports::kState);
        tr.close();
        CHECK(!tr.isConnected());
    }

    return harness::report("test_transport");
}
